### ISSUE TITLE
Double gstplayer window when using --uibc

### ISSUE THREAD

**Issue by midi1996** [albfan/miraclecast#115]
I ran miracle-sinkctl with --uibc flag but then I get two windows pooping up, one black and the other contains the viewer, also there is no command forwarding whatsoever, so I'm still just viewing. If I dont use this flag, only one window pops up as the viewer.
Here I'll attach my logs:
wifid-1 is related to sink-no-uibc
wifid-2 is related to sink-uibc
[logs.zip](https://github.com/albfan/miraclecast/files/435876/logs.zip)

also here is a small gif to show you the tests I did (sorry for the quality, I compressed it enough to make it uploadable for my slow internet speed):
http://imgur.com/a/CZYvD
Direct link:
http://i.imgur.com/0TIylLV.gifv

System information:
Ubuntu Mate 16.04
Systemd 229
miraclecast built with cmake
Qualcomm Atheros AR9285 Wireless Network Adapter (ath9k)
comamnd line ran:
`miracle-sinkctl run 3 --log-level trace --res default,CEA,00000100 --uibc`

**Comment by albfan** [comment]
That screencast is great. Can you post here the output from this command:

    $ pstree -a -p $(ps -ef | grep miracle-sinkctl | awk '{print $2}' | head -n 1)

That's to check complete command that invokes gstplayer and its parents. Install pstree if you don't have it yet

**Comment by midi1996** [comment]
Sorry for the delay.
Meanwhile I tried reinstalling ubuntu and some different computers but the same wifi card (I only have this one working, until I get my hands on a new computer, meh!)
[pstree.txt](https://github.com/albfan/miraclecast/files/445817/pstree.txt)

**Comment by albfan** [comment]
As expected there's only one gstplayer for miracle-sinkctl

Can you try

    $ ps -ef | grep miracle-sinkctl | awk '{print $2}'

That's to check there's more than one miracle-sinkctl running

**Comment by midi1996** [comment]
`ps -ef | grep miracle-sinkctl | awk '{print $2}'
28806
29358`

the first one is static and only show up when I run miraclecast-sinkctl, the second one changes each time I run this command, whether I ran miraclecast sink or not

**Comment by albfan** [comment]
Then show only

    $ ps -ef | grep miracle-sinkctl

**Comment by midi1996** [comment]
Here! apparently there is only one miracle-sinkctl running

[ps_ef.txt](https://github.com/albfan/miraclecast/files/447978/ps_ef.txt)

**Comment by midi1996** [comment]
I found something: I tried with my windows laptop and the second window didnt show up, looks like it's only related to Android devices OR with UIBC supported devices! (as I think uibc is hardware related as does M$)

**Comment by albfan** [comment]
Yes. Maybe there's a message sended two times or something like that. Anyway you can try something

When you have two gstplayers opened do this:

    $ ps -ef | grep gstplaye[r]

confirrm that it shows only two lines and they are not related to grep

then use the pid from that with pstree

    $ pstree -p <pid1>
    $ pstree -p <pid2>

Let's see from where that windows come from

Please post directly output here, it is short enough to see it directly

**Comment by albfan** [comment]
Are this solved for you then? If you can post that output we can see where it comes from. If not neccesary now just close this

**Comment by midi1996** [comment]
for `ps -ef | grep gstplayer` (for some reason zsh returned gstplaye[r] no match found)
````
root     17741 17740 20 15:55 pts/1    00:00:08 /usr/bin/python3 -u /usr/local/bin/gstplayer -d 3 -a -p 1991 -r 1920x1080
````
so I tried with bash (this time with [r]) and got the same result, basically it's only one.
Also: pstree of the gstplayer:
````
uibc-viewer(17740)───gstplayer(17741)─┬─{gstplayer}(17744)
                                      ├─{queue0:src}(17745)
                                      ├─{queue0:src}(17750)
                                      ├─{queue0:src}(17751)
                                      ├─{queue0:src}(17752)
                                      ├─{queue0:src}(17753)
                                      ├─{queue1:src}(17746)
                                      ├─{rtpjitterbuffer}(17748)
                                      ├─{threaded-ml}(17743)
                                      ├─{timer}(17747)
                                      └─{udpsrc0:src}(17749)
````

**Comment by albfan** [comment]
That numbers are piped to uibc controller. It expects a port open on source to send uibc messages. That is the cause of broken pipe too. Don't worry about that.

I think there's enough info to compare with a normal uibc screen mirroring and find the problem

Cannot test until next week anyway.

**Comment by midi1996** [comment]
Ok, I've checked my windows' laptop's capabilities and here is what sinkctl outputs:
````
wfd_client_rtp_ports
wfd_audio_codecs
wfd_video_formats
wfd_3d_video_formats
wfd_coupled_sink
wfd_display_edid
wfd_connector_type
wfd_uibc_capability            << so myeah
wfd_standby_resume_capability
wfd_content_protection
intel_sink_model_name
intel_sink_manufacturer_name
wfd_idr_request_capability
intel_sink_version
intel_sink_information
intel_lower_bandwidth
intel_interactivity_mode
intel_fast_cursor
intel_content_protection
````

 and here is from my phone:
````
wfd_video_formats
wfd_audio_codecs          << also
wfd_uibc_capability
wfd_client_rtp_ports
wfd_content_protection
wfd_sec_screensharing
````

**Comment by midi1996** [comment]
Update: Got my hands on a new computer of a friend of mine and has RTL8723BE and apparently the driver does report it supports P2P. I installed miraclecast (on Xubuntu 16.04) and tested it on that hardware, the same issue happens when casting my phone to it: double window. My windows computer doesnt give the same behavior if casted to this computer with mcast-sink. Also, I used windows 10 on this new computer and i connect it to my other computer (with ath9k AR9285 and msink) and it does connect but no video link, but if I tried the windows sink app (on the same sink computer) it does show up + i can control it with mouse/kyb/touch. Also the other way around, if I connect my old computer to this new one using windows' sink app, it doesnt let me control it (looks like its one way only, old hardware?).

**Comment by midi1996** [comment]
dude!! I tried today the wip\windows-fix branch! it fixed UIBC too (not totally, but still fixed it).
So as I said before about the wrong uibc port, this time it's getting the right one, so I still use my AR9285 and my S7 to test it out (not uibc on windows, this is windows limiting it for my hardware, blame M$, also I dont have the new laptop near me to test), I've attached my logs (miracle + android for now) and some screenshots:
[sink-log.txt](https://github.com/albfan/miraclecast/files/531311/sink-log.txt)
[wifid-log.txt](https://github.com/albfan/miraclecast/files/531312/wifid-log.txt)
![screenshot from 2016-10-15 18-10-45](https://cloud.githubusercontent.com/assets/4874726/19412065/71d7310a-9305-11e6-9ff7-e5427225ae1c.png)
![screenshot from 2016-10-15 18-11-18](https://cloud.githubusercontent.com/assets/4874726/19412066/71e2df8c-9305-11e6-9185-ec1a1f2913ce.png)
![screenshot from 2016-10-15 18-11-53](https://cloud.githubusercontent.com/assets/4874726/19412067/71ee4b38-9305-11e6-8947-ead53ff7354a.png)
![screenshot from 2016-10-15 18-12-34](https://cloud.githubusercontent.com/assets/4874726/19412068/71f9a596-9305-11e6-87fa-c48e412c9cec.png)
![screenshot from 2016-10-15 18-12-46](https://cloud.githubusercontent.com/assets/4874726/19412069/72062a78-9305-11e6-9c54-1cd8368bc53d.png)
![screenshot from 2016-10-15 18-27-39](https://cloud.githubusercontent.com/assets/4874726/19412070/7212ca30-9305-11e6-816f-bf029a2b3e82.png)
![screenshot from 2016-10-15 18-28-22](https://cloud.githubusercontent.com/assets/4874726/19412071/721c0e9c-9305-11e6-9ffa-ab2e93ef9854.png)

But there is still some issues: 
- the cast window is not re-sizable, unlike the view-only one, so im stuck with the lower part of the screen not shown
- even if I force a `CEA`/`VESA`/`HH` resolution, it's only applied when NOT using `--uibc`, otherwise it shows 1080p resolution (I tried on both laptops and I get the same result, so myeah) and I think it's either the phone who forces the resolution (given it's 2K res, and most TVs are FHD???) or miraclecast is taking my screen resolution as default resolution and forces (you know what you did, please check this one)
- no uibc with windows -> miracle (windows limitation sadly)

**Comment by albfan** [comment]
This is happening because devices not supporting UIBC is being managed incorrectly.

At this time, if you enable uibc, miraclecast tries to start miracle-uibcctl listening to UIBC port. What port? As long as there's no message from source providing such port, selected port is 0. That makes miracle-uibcctl exit with an error, that brokes the pipe on gstplayer an creates the double window.

You can reproduce it starting a non uibc session, closing actual player and launching:

    $ uibc-viewer 192.168.49.1 0 -d 3 -a -p 1991 -r 1280x720

So solution is to disable uibc if source does not support it.

I will fix this on master. And rebase WIP on top of it.

Close issue if that fixes this for you

**Comment by midi1996** [comment]
Thanks, ill try thay, just remember that UIBC on my phone started working after changes you made for windows on wip/windows-fix, i think you made sinkctl detect the port better, ill try with the master branch later. Also, the uibc capabilities are very limited (for example, as youve seen on the video link i sent on gitter, Fn keys Ctrl/Alt ESC and BckSpc doesn't work, also i cannot use the mouse wheel and swiping the pointer across the surce wont trigger a scroll but will click on the origin point you started the "gesture") Anyway this is not related to this issue, just wanted to point it out. Ill close this thread and thank you really much!

**Comment by albfan** [comment]
WIP are supposed to be rebased on top of master always.

On #88 there's a quick resume of RE to detect keys and modifiers for UIBC (There's no doc about it) Now it is more stable we can continue finding UIBC codes that emulates ESC, Home, volume, power button, mouse gestures (Now I have a laptop with touchscreen to play around), etc.

**Comment by jiahaoChen** [comment]
@albfan 
The master branch, `ctl-sink.c:256`,
`uibc_config` always gets port 0, 
looks like `uibc_config` message is wrong. 
The **uibc_config** content is `input_category_list=GENERIC;generic_cap_list=Mouse, `
It should be 
`input_category_list=GENERIC;generic_cap_list=Mouse,SingleTouch;hidc_cap_list=none;port=53056`
when I add `uibc_config = (char*)rtsp_message_get_raw(m);` after `ctl-sink.c:256`
It can get correct UIBC port.

**Comment by albfan** [comment]
On windows systems, `wfd_uibc_capability` is send with spaces, which fools miraclecast. I knew I had fixed that, but I coudn't find that commit on `wip/windows-fix`. I just forgot to push that, so thanks, here is the fix

https://github.com/albfan/miraclecast/commit/be955313524c743e32dd25c49f53c9a84f17cc74

Is that happening with a windows source or is another type of device?. I have that commit in `wip/windows-fix` because I only find that behaviour on windows. If it is another type of device I will commit that to master and rebase `wip/windows-fix`

I remember you are working with surface4, right? I have rebased that branch on top of `wip/windows-fix`.

    $ git rebase wip/windows-fix wip/surface4

Now you have all that fixes on that branch. check if you still need the ip fix and lets work on it.

Please take care of the issue you are posting on, if this issue is unrelated with your post it only add noise to it, making it less useful

Don't hesitate to open a new issue if that's the situation. I will remove this ones and post my answer there again.

**Comment by jiahaoChen** [comment]
That was happened with samsung galaxy tab s2, 
on surface4 also has the same issue,
I still test the ip fix problem, 
maybe need to open an other issue, 
surface4 is difference with laptop.

**Comment by albfan** [comment]
I found more cases related to this #183. Is related to not get the xid of the window

**Comment by albfan** [comment]
get_xid only has sense in x11 GDK backend. Nowadays Wayland seems to be the standard, so recent commits should be a workaround for this problem.

Please check if that works, I will reopen this until you confirm that resolve the issue

See you could need to execute

    xhost +si:localuser:root

to have it on boot add to init scripts

```
if [ -n "$DISPLAY" -a -z "$SSH_CLIENT" ]; then
    xhost +si:localuser:root
fi
```

**Comment by albfan** [comment]
I'm sure that's the problem

**Comment by midi1996** [comment]
Hi, I'm sorry I didnt reply earlier, I swapped my computer for work reasons. Anyway, I'm back now and I tried the latest commits and it's fixed apparently, no more double windows. However the unresizable uibc window still persists. I use `--res default,CEA,00000020` for 720p@60 with `--uibc` (same hardware btw) and it gives me an oversized window (prolly 1080p streaming), once I remove `--uibc` I can resize the window and I get 720p streaming.
